frontend/exporter/src/traits.rs

```diff
@@ -200,62 +199,90 @@ pub mod rustc {
-                    return Some(vec![]);
+        #[tracing::instrument(level = "trace", skip(s, param_env))]
+        pub fn path_to<'tcx, S: UnderOwnerState<'tcx>>(
+            starting_points: Vec<AnnotatedPredicate<'tcx>>,
+            s: &S,
+            target: PolyTraitRef<'tcx>,
+            param_env: rustc_middle::ty::ParamEnv<'tcx>,
+        ) -> Option<(Path<'tcx>, AnnotatedPredicate<'tcx>)> {
+            let tcx = s.base().tcx;
+
+            /// A candidate projects `self` along a path reaching some
+            /// predicate. A candidate is selected when its predicate
+            /// is the one expected, aka `target`.
+            #[derive(Debug)]
+            struct Candidate<'tcx> {
+                path: Path<'tcx>,
+                pred: PolyTraitPredicate<'tcx>,
+                origin: AnnotatedPredicate<'tcx>,
+            }
+
+            use std::collections::VecDeque;
+            let mut candidates: VecDeque<Candidate<'tcx>> = starting_points
+                .into_iter()
+                .map(|pred| {
+                    let clause = pred.predicate.as_trait_clause();
+                    clause.map(|clause| Candidate {
+                        path: vec![],
+                        pred: clause,
+                        origin: pred.clone(),
+                    })
+                })
+                .filter(|candidate| candidate.is_some())
+                .map(|candidate| candidate.unwrap())
+                .collect();
+
+            let target_pred = target.upcast(tcx);
+            let mut seen = std::collections::HashSet::new();
+
+            while let Some(candidate) = candidates.pop_front() {
+                {
+                    // If a predicate was already seen, we know it is
+                    // not the one we are looking for: we skip it.
+                    if seen.contains(&candidate.pred) {
+                        continue;
+                    }
+                    seen.insert(candidate.pred.clone());
+                }
+
+                // if the candidate equals the target, let's return its path
+                if predicate_equality(candidate.pred.upcast(tcx), target_pred, param_env, s) {
+                    return Some((candidate.path, candidate.origin));
                 }
 
-                let recurse = |p: Self| {
-                    if p == self {
-                        return None;
+                // otherwise, we add to the queue all paths reachable from the candidate
+                for (index, parent_pred) in candidate.pred.parents_trait_predicates(s) {
+                    let mut path = candidate.path.clone();
+                    path.push(PathChunk::Parent {
+                        predicate: parent_pred.clone(),
+                        index,
+                    });
+                    candidates.push_back(Candidate {
+                        pred: parent_pred.clone(),
+                        path,
+                        origin: candidate.origin.clone(),
```

Comment:
I think we can make `AnnotatedPredicate` copy. Can you add `Copy` to the derives of `rustc_utils::AnnotatedPredicate`, and then remove `clone`s?